app/lib/shared/deps_graph.dart

```diff
@@ -0,0 +1,218 @@
+          ..order('created');
+        await for (PackageVersion pv in query.run()) {
+          // Emit a set of packages depending on this one.
+          _addPackageVersion(pv);
+          yield _affectedPackages(pv.package);
+
+          _lastTs = pv.created;
+        }
+      } catch (e, s) {
+        _logger.severe(e, s);
+      }
+      await new Future.delayed(pollingInterval);
+    }
+  }
+
+  void _addPackageVersion(PackageVersion pv) {
+    final String package = pv.package;
+    final Map pubspec = pv.pubspec.asJson;
+
+    final Set<String> depsSet = new Set<String>();
+    final normalDeps = pubspec['dependencies'];
+    if (normalDeps is Map) {
+      depsSet.addAll(normalDeps.keys);
+    }
+
+    final Set<String> devDepsSet = new Set<String>();
+    final devDeps = pubspec['dev_dependencies'];
+    if (devDeps is Map) {
+      devDepsSet.addAll(devDeps.keys);
+    }
+
+    // First we add the [package] together with the dependencies /
+    // dev_dependencies to the graph.  This will update the graph transitively,
+    // thereby calculating new users of the [package].
+    _add(package, depsSet, devDepsSet);
+  }
+
+  Set<String> _affectedPackages(String package) {
+    // Due to the constraints in the new [pubspec] it might cause a number of
+    // packages to get new transitive dependencies during a `pub get` and
+    // therefore might cause new analysis results.
+    //
+    // So we trigger all packages that depend (directly or indirectly) on
+    // [package].
+    final Set<String> transitiveUsers =
+        reverseDeps.transitiveNodes(canonicalize(package));
+
+    // We filter out all dev package usages and trigger an analysis for them
+    // (the dev packages are a superset of the normal packages).
+    final Set<String> all = transitiveUsers
+        .where((String p) => p.startsWith(devPrefix))
+        .map((String p) => p.substring(devPrefix.length)).toSet();
+    return all;
+  }
+
+  void _add(String package, Set<String> deps, Set<String> devDeps) {
+    final canonicalizedPackage = canonicalize(package);
+    final canonicalizedDevPackage = canonicalize(_devPackage(package));
+
+    final canonicalizedDeps = deps.map(canonicalize).toSet();
+    final canonicalizedDevDeps = devDeps.map(canonicalize).toSet();
+
+    //normalDeps.addAll(canonicalizedPackage, canonicalizedDeps);
+    //normalDeps.addAll(canonicalizedDevPackage, canonicalizedDeps);
+    for (final String dep in canonicalizedDeps) {
+      reverseDeps.add(dep, canonicalizedPackage);
+      reverseDeps.add(dep, canonicalizedDevPackage);
+    }
+
+    //normalDeps.addAll(canonicalizedDevPackage, canonicalizedDevDeps);
+    for (final String dep in canonicalizedDevDeps) {
+      reverseDeps.add(dep, canonicalizedDevPackage);
+    }
+  }
+
+  // We use a synthetic package to handle weak dependencies / dev
+  // dependencies (no package is depending on this one).
+  String _devPackage(String package) => '$devPrefix$package';
```

Comment:
I'm not sure that it is worth to differentiate between dev and direct dependencies in our case. I've had written a package where I put a dependency as dev, because the idea was to have a common interface for various databases, and put db-specific implementations alongside. Not a great design, but it happens, I've seen in another project too. Better to be more pessimistic and treat every dependency the same way.

Reply:
With the recent discussion on [dependencies](https://github.com/dart-lang/pub-dartlang-dart/issues/441) I'm wondering whether this is needed at all.

Reply:
I would prefer to submit it with the support, then it's at least in the git history in case we change our minds later on.

We can remove it, yes. It depends what `package:pana` does. If it only looks into the `lib` folder, then I would be fine with it. If it also looks into other folders, I guess we should keep it.  (Actually I have no idea what the policy about the `bin` folder is: It contains entry points which are exposed like the `lib` folder. So I assume if one uses a package in `bin` then it has to go into the normal `dependencies` section???)

Reply:
sgtm, but you need to fix the tests too (mocks require the new signatures)
`lib` and `bin` is treated the same way wrt. dependencies, because you could run the script from the `bin/` directory with `pub run`.